File: apps/dashboard/src/lib/fleet.ts
/**
 * Fleet Option A — static sub-budget partitioning (mirrors mintry.core.fleet).
 *
 * sum(shares) ≤ total_usd. Each agent later enforces only its local max_usd.
 */

export const MIN_SHARE_USD = 0.01;
const EPSILON = 1e-9;

export type FleetPartitionError = { ok: false; error: string };
export type FleetPartitionPlan = {
  ok: true;
  fleet_id: string;
  total_usd: number;
  partitions: Record<string, number>;
  allocated_usd: number;
  unallocated_usd: number;
};

export function validatePartitions(
  totalUsd: number,
  partitions: Record<string, number>,
  opts?: { fleetId?: string; minShareUsd?: number }
): FleetPartitionPlan | FleetPartitionError {
  const minShare = opts?.minShareUsd ?? MIN_SHARE_USD;

  if (typeof totalUsd !== "number" || !Number.isFinite(totalUsd)) {
    return { ok: false, error: "total_usd must be a finite number" };
  }
  if (totalUsd < minShare) {
    return {
      ok: false,
      error: `total_usd must be >= ${minShare.toFixed(2)} (got ${totalUsd})`,
    };
  }

  if (!partitions || typeof partitions !== "object" || Array.isArray(partitions)) {
    return {
      ok: false,
      error: "partitions must be an object map of agent_id → share_usd",
    };
  }

  const entries = Object.entries(partitions);
  if (entries.length === 0) {
    return { ok: false, error: "partitions must include at least one agent" };
  }

  const normalized: Record<string, number> = {};
  for (const [rawAgent, rawShare] of entries) {
    const agentId = String(rawAgent ?? "").trim();
    if (!agentId) {
      return { ok: false, error: "every partition key must be a non-empty agent_id" };
    }
    if (typeof rawShare !== "number" || !Number.isFinite(rawShare)) {
      return {
        ok: false,
        error: `share for agent '${agentId}' must be a finite number`,
      };
    }
    if (rawShare < minShare) {
      return {
        ok: false,
        error: `share for agent '${agentId}' must be >= ${minShare.toFixed(2)} (got ${rawShare})`,
      };
    }
    if (agentId in normalized) {
      return { ok: false, error: `duplicate agent_id in partitions: '${agentId}'` };
    }
    normalized[agentId] = Math.round(rawShare * 1e6) / 1e6;
  }

  const allocated =
    Math.round(Object.values(normalized).reduce((a, b) => a + b, 0) * 1e6) / 1e6;
  if (allocated > totalUsd + EPSILON) {
    return {
      ok: false,
      error: `fleet oversubscribed: allocated $${allocated.toFixed(4)} exceeds total $${totalUsd.toFixed(4)}`,
    };
  }

  const total = Math.round(totalUsd * 1e6) / 1e6;
  return {
    ok: true,
    fleet_id: (opts?.fleetId || "").trim(),
    total_usd: total,
    partitions: normalized,
    allocated_usd: allocated,
    unallocated_usd: Math.round(Math.max(total - allocated, 0) * 1e6) / 1e6,
  };
}

export function mandateRuleForShare(
  shareUsd: number,
  opts?: { fleetId?: string; fleetTotalUsd?: number }
): Record<string, unknown> {
  const rule: Record<string, unknown> = {
    max_usd: shareUsd,
    allow: true,
  };
  if (opts?.fleetId) {
    rule.fleet_id = opts.fleetId;
  }
  if (opts?.fleetTotalUsd !== undefined) {
    rule.fleet_total_usd = opts.fleetTotalUsd;
  }
  return rule;
}

File: src/mintry/core/fleet.py
"""Fleet Option A — static sub-budget partitioning.

Authors split a fleet total into per-agent shares at the control plane.
Each agent then enforces only its local ``max_usd`` (Principle 3 + 6).
No shared counter and no network I/O on the authorize hot path.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping


MIN_SHARE_USD = 0.01
_EPSILON = 1e-9


@dataclass(frozen=True)
class FleetPartitionError:
    """Validation failure for a proposed fleet partition."""

    message: str

    def __str__(self) -> str:
        return self.message


@dataclass(frozen=True)
class FleetPartitionPlan:
    """Validated static partition ready to author as per-agent policies."""

    fleet_id: str
    total_usd: float
    partitions: dict[str, float]
    allocated_usd: float
    unallocated_usd: float


def validate_partitions(
    total_usd: float,
    partitions: Mapping[str, float],
    *,
    fleet_id: str = "",
    min_share_usd: float = MIN_SHARE_USD,
) -> FleetPartitionPlan | FleetPartitionError:
    """Validate Option A static partitions.

    Accepted when:
    - ``total_usd`` is finite and >= ``min_share_usd``
    - ``partitions`` is a non-empty map of agent_id → share
    - every agent_id is a non-empty string
    - every share is finite and >= ``min_share_usd``
    - ``sum(shares) <= total_usd`` (undersubscribe keeps unallocated headroom)

    Rejected when the fleet is oversubscribed (``sum > total``).
    """
    if not isinstance(total_usd, (int, float)) or isinstance(total_usd, bool):
        return FleetPartitionError("total_usd must be a number")
    total = float(total_usd)
    if total != total or total in (float("inf"), float("-inf")):  # NaN / inf
        return FleetPartitionError("total_usd must be finite")
    if total < min_share_usd:
        return FleetPartitionError(
            f"total_usd must be >= {min_share_usd:.2f} (got {total})"
        )

    if not isinstance(partitions, Mapping) or isinstance(partitions, (str, bytes)):
        return FleetPartitionError("partitions must be an object map of agent_id → share_usd")
    if not partitions:
        return FleetPartitionError("partitions must include at least one agent")

    normalized: dict[str, float] = {}
    for raw_agent, raw_share in partitions.items():
        agent_id = str(raw_agent).strip() if raw_agent is not None else ""
        if not agent_id:
            return FleetPartitionError("every partition key must be a non-empty agent_id")
        if not isinstance(raw_share, (int, float)) or isinstance(raw_share, bool):
            return FleetPartitionError(f"share for agent {agent_id!r} must be a number")
        share = float(raw_share)
        if share != share or share in (float("inf"), float("-inf")):
            return FleetPartitionError(f"share for agent {agent_id!r} must be finite")
        if share < min_share_usd:
            return FleetPartitionError(
                f"share for agent {agent_id!r} must be >= {min_share_usd:.2f} (got {share})"
            )
        if agent_id in normalized:
            return FleetPartitionError(f"duplicate agent_id in partitions: {agent_id!r}")
        normalized[agent_id] = round(share, 6)

    allocated = round(sum(normalized.values()), 6)
    if allocated > total + _EPSILON:
        return FleetPartitionError(
            f"fleet oversubscribed: allocated ${allocated:.4f} exceeds total ${total:.4f}"
        )

    fid = (fleet_id or "").strip()
    return FleetPartitionPlan(
        fleet_id=fid,
        total_usd=round(total, 6),
        partitions=normalized,
        allocated_usd=allocated,
        unallocated_usd=round(max(total - allocated, 0.0), 6),
    )


def mandate_rule_for_share(
    share_usd: float,
    *,
    fleet_id: str = "",
    fleet_total_usd: float | None = None,
) -> dict:
    """Build the flat mandate rule embedded in a per-agent policy bundle.

    Extra fleet metadata is informational for audit/UI. The hot path only
    consumes ``max_usd`` / ``allow`` / ``expires_at``.
    """
    rule: dict = {"max_usd": float(share_usd), "allow": True}
    if fleet_id:
        rule["fleet_id"] = fleet_id
    if fleet_total_usd is not None:
        rule["fleet_total_usd"] = float(fleet_total_usd)
    return rule
